### PR TITLE
Agent actions token support

### PR DESCRIPTION
**Pull request by aleksmaus**
## What does this PR do?

Implements the actions token exchange.
This includes the stub for the agent actions that just logs the action received (DEBUG log level). 
The actions will need need to be wired in further to properly pass them to the apps/beats and receive the result back. @blakerouse let me know if you are going to handle that or if I should dig further.

The flow of the action token exchange is the following.
1. Fleet Server sends the ack_token string with that actions to the agent that serves as a "mark" of the latest action received by the agent. 
Screenshot of the app/beat action logged from the stub:
Screen Shot 2021-01-12 at 10 33 33 AM

2. Agent persists the ack_token value (currently in a separate file action_ack_token.yml 
Screen Shot 2021-01-12 at 10 51 57 AM

@blakerouse let me know if there is a better place for that, since I'm just getting familiar with the agent code.

3. Agent sends the ack_token string with the next check-in request to Fleet Server.
4. Fleet Server decodes/translates the ack_token string into the action doc sequence number and updates the agent record `action_seq_no`

Screen Shot 2021-01-12 at 10 52 08 AM

This way the fleet server tracks the latest action received by the agent.

If the ack_token is not present in the check in payload the value that is stored with the agent record is used.

## Why is it important?

This is needed to support the new Fleet Server agent actions handling on the agent side. Without this change the new action document in the .fleet-actions will cause the agent to go into a loop of check-ins and receiving the same action over and over, since there will be no indication that the agent action was received. 

@blakerouse One question about the persisted ack_token on the agent side. We probably should remove the file every time the agent enrolls, since it creates the new agent record for the fleet. Thoughts?

### CODE REVIEW

**Review by james-elastic — APPROVED** [review]

File: x-pack/elastic-agent/pkg/agent/application/info/agent_id.go [hunk 1]
```diff
@@ -43,11 +44,16 @@ func AgentConfigFile() string {
 	return filepath.Join(paths.Config(), defaultAgentConfigFile)
 }
 
-// AgentActionStoreFile is the file that will contains the action that can be replayed after restart.
+// AgentActionStoreFile is the file that contains the action that can be replayed after restart.
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/agent/application/info/agent_id.go:47]: Being that we already have an `action_store.yml` why not place the action token inside of this file? Why the need to seperate it into its own file?

Comment (aleksmaus) [reply x-pack/elastic-agent/pkg/agent/application/info/agent_id.go:47]: can change. was not sure what's the right/established pattern.

Comment (aleksmaus) [reply x-pack/elastic-agent/pkg/agent/application/info/agent_id.go:47]: is there any case where the `agent_store.yml` could get completely overwritten, thus loosing the marker?

Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/application/info/agent_id.go:47]: i think that would be the correct place

It should only be overwritten on a re-enroll, which is what should happen in the token case. The code paths already handle that, so its the best place for it.

File: x-pack/elastic-agent/pkg/agent/application/acktoken_store.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+func (s *ackTokenStore) GetToken() string {
+	mx.RLock()
+	ats := atsCached
+	mx.RUnlock()
+	return ats.AckToken
+}
+
+type ackTokenSerializer struct {
+	AckToken string `yaml:"ack_token"`
```
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/application/acktoken_store.go:85]: Maybe worth to make the mutex directly part of this struct? Then you could use `atsCached.Lock()` which would make the code likely more readable.

File: x-pack/elastic-agent/pkg/fleetapi/checkin_cmd.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ const checkingPath = "/api/fleet/agents/%s/checkin"
 // CheckinRequest consists of multiple events reported to fleet ui.
 type CheckinRequest struct {
 	Status   string              `json:"status"`
+	AckToken string              `json:"ack_token"`
```
Comment (nchaulet) [inline x-pack/elastic-agent/pkg/fleetapi/checkin_cmd.go:25]: Should we omit if empty? otherwise we probably need to add the property to Kibana

Comment (aleksmaus) [reply x-pack/elastic-agent/pkg/fleetapi/checkin_cmd.go:25]: Sure. Will update.

**Comment by aleksmaus** [comment]
Discussed with @blakerouse, going to consolidate the ```action_store.yml``` and ```action_ack_token.yml``` files into more generic ```state.yml``` storage format that can be used for both and extended further. 
The agent will handle the migration on first start.

**Review by blakerouse — APPROVED** [review]
Looks really good, well tested! Please backport this to 7.x.

**Pull request merged**